internal/processor/stream/processor.go

```diff
@@ -256,10 +265,13 @@ func (p *Processor) HandleStream(
 			// a sync.Pool for creating batches, and having the publisher (terminal processor)
 			// release batches back into the pool.
 			if err := processor.ProcessBatch(ctx, &batch); err != nil {
+				releaseBatch()
 				return err
 			}
 			result.AddAccepted(len(batch))
 		}
+		releaseBatch()
```

Comment:
I don't think this is safe. As mentioned above, ProcessBatch takes ownership, so it should be the one to release the batch. I'd have to audit the codebase to be sure, but IIRC there's just one processor where this matters: the libbeat publisher. See this code, which processes batches asynchronously:

https://github.com/elastic/apm-server/blob/a4657d87dc0f336555d829a67b4aa04a8c800fa5/internal/publish/pub.go#L138-L174

Reply:
Good catch!

I reviewed the codebase, it seems that's the only processor which expects the batch slice to be valid after the method has returned. I've pushed a fix!